### PR TITLE
TextAlignment property added to NumberBox

### LINKED ISSUE #1722: Proposal: Add TextAlignment property to NumberBox
# Proposal: Add TextAlignment property to NumberBox

## Summary
It would be nice if NumberBox also had a TextAlignment property. 

In this screenshot the NumberBoxes have Right TextAlignment 
image

## Rationale

* Make it consistent with TextBox and Excel
* Creating a Template for the NumberBox with only this extra property is a lot of work. Do-able but not nice.

```
<TextBox x:Name="InputBox"
                                     Grid.Column="0"
                                     TextAlignment="Right"
                                     Header="{TemplateBinding Header}"
                                     HeaderTemplate="{TemplateBinding HeaderTemplate}"
                                     PreventKeyboardDisplayOnProgrammaticFocus="{TemplateBinding PreventKeyboardDisplayOnProgrammaticFocus}"
                                     PlaceholderText="{TemplateBinding PlaceholderText}"
                                     SelectionHighlightColor="{TemplateBinding SelectionHighlightColor}"
                                     TextReadingOrder="{TemplateBinding TextReadingOrder}" />
```

### PR DESCRIPTION
**Pull request by sonnemaf**
The NumberBox was missing a TextAlignment property. Now it has one.

## Description
I added the TextAlignment dependency property to the NumberBox. It needs testing and documentation. Don't know how to do this. Sorry.

## Motivation and Context
Closes #1722

## How Has This Been Tested?
Added two NumberBoxes to the test page with Center and Right TextAlignment.

### CODE REVIEW

File: dev/NumberBox/NumberBox.xaml [hunk 1]
```diff
@@ -1,14 +1,13 @@
-﻿<!-- Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT License. See LICENSE in the project root for license information. -->
-<ResourceDictionary
```
Comment (chingucoding) [inline dev/NumberBox/NumberBox.xaml:2]: Almost all of those changes are just formatting. Can you please revert the formatting changes? 
With all these changes it is hard to review the actual new feature in question.

File: dev/Generated/NumberBox.properties.h [hunk 2]
```diff
@@ -82,6 +85,7 @@ class NumberBoxProperties
     static winrt::DependencyProperty SmallChangeProperty() { return s_SmallChangeProperty; }
     static winrt::DependencyProperty SpinButtonPlacementModeProperty() { return s_SpinButtonPlacementModeProperty; }
     static winrt::DependencyProperty TextProperty() { return s_TextProperty; }
+    static winrt::DependencyProperty TextAlignmentProperty() { return s_TextAlignmentProperty; }
```
Comment (chingucoding) [inline dev/Generated/NumberBox.properties.h:88]: I think this would technically go through API review first. @MikeHillberg Should this be marked as preview until then?

File: dev/NumberBox/NumberBox.idl [hunk 3]
```diff
@@ -60,6 +60,7 @@ unsealed runtimeclass NumberBox : Windows.UI.Xaml.Controls.Control
     String PlaceholderText;
     Windows.UI.Xaml.Controls.Primitives.FlyoutBase SelectionFlyout;
     Windows.UI.Xaml.Media.SolidColorBrush SelectionHighlightColor;
+    Windows.UI.Xaml.TextAlignment TextAlignment;
```
Comment (chingucoding) [inline dev/NumberBox/NumberBox.idl:63]: Posting previous comment here as this is more suitable in the IDL file.

I think this would technically go through API review first. @MikeHillberg Should this be marked as preview until then?

**Comment by chingucoding** [comment]
You can add a test NumberBox which uses the TextAlignment property on the test page inside the NumberBox_TestUI project(which is under /dev/NumberBox), to verify that the changes work correctly.

What kind of information would have helped you [on the source code structure documentation](https://github.com/microsoft/microsoft-ui-xaml/blob/master/docs/source_code_structure.md) to make it easier understanding the source code?

File: dev/NumberBox/NumberBox.xaml (hunk 1 above)
Comment (sonnemaf) [reply dev/NumberBox/NumberBox.xaml]: I just reverted the changes. I had XAML Styler running which moved things around. Sorry.

**Comment by sonnemaf** [comment]
> You can add a test NumberBox which uses the TextAlignment property on the test page inside the NumberBox_TestUI project(which is under /dev/NumberBox), to verify that the changes work correctly.
> 
> @sonnemaf What kind of information would have helped you [on the source code structure documentation](https://github.com/microsoft/microsoft-ui-xaml/blob/master/docs/source_code_structure.md) to make it easier understanding the source code?

The source code is quite clear (although I have no real experience in C++. I just couldn't find the test app. I found the MUXControlsTestApp app and I can now test the app.

**Comment by Felix-Dev** [comment]
For the API review: @MikeHillberg and I are currently looking at an [apparent mismatch](https://github.com/microsoft/microsoft-ui-xaml/issues/2896) between the documentation and the API behavior regarding the `TextBox.HorizontalTextAlignment` and `TextBox.TextAlignment` APIs. This is relevant here because the naming of this API perhaps should be changed to `NumberBox.HorizontalTextAlignment` instead (depending on how the mismatch will be resolved).

File: dev/NumberBox/NumberBox.idl (hunk 3 above)
Comment (Felix-Dev) [reply dev/NumberBox/NumberBox.idl:63]: Yes, this should be reviewed and marked as preview because of an [apparent mismatch](https://github.com/microsoft/microsoft-ui-xaml/issues/2896) between the documentation and the API behavior regarding the TextBox.HorizontalTextAlignment and TextBox.TextAlignment APIs.

**Comment by MikeHillberg** [comment]
A spec for this is a good place to have a discussion about those two alignment properties. Also the [InputScope property](https://github.com/microsoft/microsoft-ui-xaml/pull/2605).

**Comment by StephenLPeters** [comment]
@SavoySchuler FYI

File: dev/NumberBox/TestUI/NumberBoxPage.xaml [hunk 4]
```diff
@@ -1,154 +1,252 @@
...
+                    <!--  TextAlignment  -->
+                    <controls:NumberBox Header="Right TextAlignment"
+                                        TextAlignment="Right"
+                                        Value="123" />
+                    <controls:NumberBox Header="Center TextAlignment"
+                                        TextAlignment="Center"
+                                        Value="456" />
+
+
+                    <controls:NumberBox x:Name="TestNumberBox"
```
Comment (StephenLPeters) [inline dev/NumberBox/TestUI/NumberBoxPage.xaml:165]: Can you please add a Combo box which has the text alignment options as combo box items and selecting one of those will change the TestNumberBox's TextAlignment to the chosen value.

File: dev/NumberBox/NumberBox.idl (hunk 3 above)
Comment (StephenLPeters) [reply dev/NumberBox/NumberBox.idl:63]: @sonnemaf until this has an approved spec, you'll need to mark the API as preview, examples can been seen in some other IDL files in the repo

Comment (Felix-Dev) [reply dev/NumberBox/NumberBox.idl:63]: @sonnemaf For example [here](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Repeater/ItemsRepeater.idl#L439).

File: dev/NumberBox/TestUI/NumberBoxPage.xaml (hunk 4 above)
Comment (sonnemaf) [reply dev/NumberBox/TestUI/NumberBoxPage.xaml:166]: Great suggestion. Done!

image

File: dev/NumberBox/NumberBox.idl (hunk 3 above)
Comment (sonnemaf) [reply dev/NumberBox/NumberBox.idl]: Done!

File: dev/NumberBox/NumberBox.idl [hunk 5]
```diff
@@ -60,7 +60,12 @@ unsealed runtimeclass NumberBox : Windows.UI.Xaml.Controls.Control
     String PlaceholderText;
     Windows.UI.Xaml.Controls.Primitives.FlyoutBase SelectionFlyout;
     Windows.UI.Xaml.Media.SolidColorBrush SelectionHighlightColor;
-    Windows.UI.Xaml.TextAlignment TextAlignment;
+
+    [WUXC_VERSION_PREVIEW]
```
Comment (Felix-Dev) [inline dev/NumberBox/NumberBox.idl:64]: You also need to include `static Windows.UI.Xaml.DependencyProperty TextAlignmentProperty{ get; };` here.

File: dev/Generated/NumberBox.properties.cpp [hunk 6]
```diff
@@ -226,7 +226,7 @@ void NumberBoxProperties::EnsureProperties()
                 winrt::name_of<winrt::TextAlignment>(),
                 winrt::name_of<winrt::NumberBox>(),
                 false /* isAttached */,
-                ValueHelper<winrt::TextAlignment>::BoxValueIfNecessary(winrt::TextAlignment::Left),
```
Comment (Felix-Dev) [inline dev/Generated/NumberBox.properties.cpp]: Why did you remove the default value? TextBox.TextAlignment/HorizontalTextAlignment uses TextAlignment::Left as their default value and we should pick the same default here.

To add a default value, you add the default value in the NumberBox .idl file by adding `[MUX_DEFAULT_VALUE("winrt::TextAlignment::Left")]` above the TextAlignment entry (you can find examples in the .idl file).

File: dev/NumberBox/TestUI/NumberBoxPage.xaml [hunk 7]
```diff
@@ -1,154 +1,253 @@
...
-    x:Name="NumberBoxTestPage"
-    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
-    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
-    xmlns:local="using:MUXControlsTestApp"
-    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
-    xmlns:controls="using:Microsoft.UI.Xaml.Controls"
-    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
-    xmlns:contract5Present="http://schemas.microsoft.com/winfx/2006/xaml/presentation?IsApiContractPresent(Windows.Foundation.UniversalApiContract,5)"
-    mc:Ignorable="d">
+﻿<!--  Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT License. See LICENSE in the project root for license information.  -->
```
Comment (chingucoding) [inline dev/NumberBox/TestUI/NumberBoxPage.xaml:1]: A lot of these changes are only formatting, please revert them.

File: dev/Generated/NumberBox.properties.cpp (hunk 6 above)
Comment (chingucoding) [reply dev/Generated/NumberBox.properties.cpp]: Agree with @Felix-Dev here, please set the default value to `TextAlignment::Left`.

Comment (sonnemaf) [reply dev/Generated/NumberBox.properties.cpp]: Don't know what happened with the Default. It is now back to Left. Sorry...

File: dev/NumberBox/NumberBox.idl (hunk 5 above)
Comment (sonnemaf) [reply dev/NumberBox/NumberBox.idl:64]: Done, makes sense.

File: dev/Generated/NumberBox.properties.cpp (hunk 6 above)
Comment (Felix-Dev) [reply dev/Generated/NumberBox.properties.cpp]: @sonnemaf It appears to me you have manually edited the NumberBox.properties.cpp file to include the default value in commit https://github.com/microsoft/microsoft-ui-xaml/pull/2901/commits/dc9cef730e98f749bc7148b8f4cd44bec22de93d. Since that file is auto-generated (for example whenever a new API is added to the NumberBox.idl file) manually added code *will not* persist here. In fact, I wouldn't be surprised if that was the reason the default value was removed in a previous commit (the file was probably automatically generated again and you committed that new version). 

Instead, to specify a persisting default value for a public API, please add it to the .idl file like so:
```
[WUXC_VERSION_PREVIEW]
{
      [MUX_DEFAULT_VALUE("winrt::TextAlignment::Left")]
      Windows.UI.Xaml.TextAlignment TextAlignment;
}
```

This will generate the correct code whenever a new version of that file is automatically created.

Comment (sonnemaf) [reply dev/Generated/NumberBox.properties.cpp]: Yes I was editing the NumberBox.properties.cpp manually. I fixed it now using the MUX_DEFAULT_VALUE attribute in the .idl file. C++ is totally new for me. Thanks for your help.

File: dev/NumberBox/TestUI/NumberBoxPage.xaml (hunk 7 above)
Comment (sonnemaf) [reply dev/NumberBox/TestUI/NumberBoxPage.xaml:1]: Sorry, XAML Styler was interfering again. Disabled it for now. 

I have reverted the changes. I should be ok again.

Comment (chingucoding) [reply dev/NumberBox/TestUI/NumberBoxPage.xaml:1]: I think you forgot to push that commit 😅

Comment (StephenLPeters) [reply dev/NumberBox/TestUI/NumberBoxPage.xaml:1]: I'd also prefer the styling update to be a separate commit

---
In reply to: [456551014](https://github.com/microsoft/microsoft-ui-xaml/pull/2901#discussion_r456551014) [](ancestors = 456551014)

Comment (sonnemaf) [reply dev/NumberBox/TestUI/NumberBoxPage.xaml]: I forgot the Push 😒 and made now two separate commits 😊.

**Comment by StephenLPeters** [comment]
@ranjeshj and @SavoySchuler I think the change looks good assuming the API does.

**Comment by Felix-Dev** [comment]
Should we also add an API test here?

**Comment by ranjeshj** [comment]
> @ranjeshj and @SavoySchuler I think the change looks good assuming the API does.

I think its fine to merge as a preview API.

**Comment by StephenLPeters** [comment]
> Should we also add an API test here? (based on the current implementation: verifying that setting this API sets the corresponding API on the TextBox - for which I assume tests exist to verify its correctness.)

I think this is a valid consideration.  @sonnemaf do you have time to add a test?

**Comment by StephenLPeters** [comment]
> > Should we also add an API test here? (based on the current implementation: verifying that setting this API sets the corresponding API on the TextBox - for which I assume tests exist to verify its correctness.)
> 
> I think this is a valid consideration. @sonnemaf do you have time to add a test?

@sonnemaf Do you need help writing a tests?
